diesel/src/query_builder/upsert/on_conflict_actions.rs

```diff
@@ -1,14 +1,23 @@
-use crate::backend::{sql_dialect, Backend};
+use std::marker::PhantomData;
+
+use crate::backend::sql_dialect::on_conflict_clause;
+use crate::backend::Backend;
 use crate::expression::{AppearsOnTable, Expression};
 use crate::query_builder::*;
 use crate::query_source::*;
 use crate::result::QueryResult;
 
 #[doc(hidden)]
 #[derive(Debug, Clone, Copy, QueryId)]
-pub struct DoNothing;
+pub struct DoNothing<T>(PhantomData<T>);
+
+impl<T> DoNothing<T> {
+    pub(crate) fn new() -> Self {
+        Self(PhantomData)
+    }
+}
 
-impl<DB> QueryFragment<DB> for DoNothing
+impl<T, DB> QueryFragment<DB> for DoNothing<T>
```

Comment:
For this and the rest of the implementations in this file... why the need to introduce the new generic parameter(s)?

Reply:
Because we need to know the table type in the `QueryFragment` impl for `DoNothing (and `DoUpdate`) to emit the `id = id` workaround there. (Specifically we need to know the name of the primary key there.)

Reply:
Thank you this was really helpful!